### PR TITLE
initial attempt to refactor paragraph

### PR DESCRIPTION
**Pull request by heruka-urgyen**
1. add lodash
2. use `controller as` pattern instead of scope
3. major internal changes in controller. not done yet. but let's start from here

### CODE REVIEW

File: zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js [hunk 1]
```diff
@@ -24,50 +24,79 @@
...
+    $rootScope, 
+    $route, 
+    $window, 
+    // $routeParams, 
+      // $location, 
+    $timeout, 
+    $compile
+  ) {
+
+  var vm = this;
```
Comment (corneadoug) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js:40]: Why vm?
I'm familiar with var self = this, but not sure of vm meaning

Comment (heruka-urgyen) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js:40]: vm stands for "view-model"

File: zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js [hunk 2]
```diff
@@ -1057,563 +966,288 @@ angular.module('zeppelinWebApp')
...
+  vm.toggleOutput = toggleOutput
+  vm.toggleGraphOption = toggleGraphOption
+  vm.changeColWidth = changeColWidth
+  vm.setColumnWidthClass = setColumnWidthClass
+  vm.showTitle = showTitle
+  vm.hideTitle = hideTitle
+  vm.setTitle = setTitle
+  vm.toggleEditor = toggleEditor
+  vm.removeParagraph = removeParagraph
+  vm.moveUp = moveUp
```
Comment (corneadoug) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js:1245]: What is the difference between:

var my_function = function() {};

and

function my_function() {}
vm.my_function = my_function?

Comment (heruka-urgyen) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js:1245]: `function my_function() {}` is hoisted so it's more reliable than `var my_function = function() {};`. those are controller-only functions.
`vm.my_function = my_function?` is available in the view.

Comment (corneadoug) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js:1245]: Didn't know you could access functions outside of the controller scope from the view

Comment (heruka-urgyen) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js:1245]: that's the whole point of `controller as` pattern.

**Comment by corneadoug** [comment]
Hey @felizbear, while testing your branch, I discovered a couple of bugs of Zeppelin (that were there before).

Since it looks complicated to work on both PR at the same time, I'm thinking about merging yours in mine, and then make test + bug fix in the original PR, what do you think?

**Comment by heruka-urgyen** [comment]
It's fine with me. Let's do it

**Pull request closed without merging**